### DISCUSSION TITLE
Is it possible to create a tree-shakeable library with Parcel?

### DISCUSSION THREAD

**Discussion by cseas** [parcel-bundler/parcel#8619] [Q&A]
I couldn't find any reference in the [docs][1] regarding this. Use-case is to ship a TypeScript component library to npm that users can consume and tree-shake themselves (any unnecessary components should be removed from their bundle).

So far I've found Rollup is the only tool that's able to do this with `--preserveModules`. Webpack has an [ongoing issue][3] with [limited experimental support][2] for it.

If I'm bundling my library with, say, 2 components, A & B. And my consumer is only using component A, how do I make sure they should be able to tree-shake B?

[1]: https://parceljs.org/getting-started/library/
[2]: https://webpack.js.org/configuration/output/#type-module
[3]: https://github.com/webpack/webpack/issues/2933

**Comment by Nargonath** [comment]
I have the same use case. @cseas did you find a solution? I tried with Rollup but to no avail. I have errors when consuming the result build in the consumer application and haven't found a way to fix them yet.

**Reply by cseas** [reply]
Hey, I eventually gave up on Parcel due to lack of documentation at the time. I found Rollup to be fairly straightforward to setup. I specified every file as an entry point as mentioned in the [Rollup docs][1] to preserve the original tree I wanted to ship.

If you need a reference, my package is `hazel-ui` on npm and the relevant part of Rollup config that preserves the original tree structure is [here][2].

[1]: https://rollupjs.org/configuration-options/#input
[2]: https://github.com/hazel-ui/hazel-ui/blob/main/rollup.config.ts#L19-L31

**Reply by Nargonath** [reply]
Cool, thank you. I'll have a look at it.

**Reply by Nargonath** [reply]
@cseas Thank you very much for your guidance. It helped fixed my current problem. However now that I rewrote the path for the `input` the paths don't match anymore with the sourceMap. I'm looking into `sourcemapPathTransform` but it seems that the function is not called for my component files.
